Add contains option to checks.

MFraters opened the pull request:
## Description
This is an implementation of what was suggested in #619. It implements a `Contains` class for the `CHECK_THROWS_WITH` (and checks in general). This allows for checking whether the throw message contains a certain string, similar to what Catch2 does. 

I first tried the union approach as suggested in the related issue, but I ran in a host of issues, including that due to non-trivial members, the default constructors and operators where deleted. I found especially the copy constructor/operator= hard to replace, since it requires knowledge about what member of the union is being used. I now use a bool to check whether it is a contains or not. This approach seems easier to me, and I can't see significant downsides of this implementation, but I am open to further suggestions.

I have not implemented test yet, because I would first like to know if this approach is acceptable, but it works well as a replacement for catch2 on my code base. I tried to format the cod with `clang-format -style=file -i doctest/parts/doctest*`, but it also changes the format of lines I haven't changed in this pull request.

## GitHub Issues
closes  #619.

Saalvage commented on doctest/parts/doctest_fwd.h:
```diff
@@ -576,12 +578,21 @@ class DOCTEST_INTERFACE String
 
     int compare(const char* other, bool no_case = false) const;
     int compare(const String& other, bool no_case = false) const;
+    const char *compare(const Contains& other) const;
```
Why does this return a `const char*`? Imo this should be a method on `Contains` (e.g.: `Contains::checkWith(String)`) that simply returns a bool. This will also make the `Contains` forward declaration superfluous and make the dependency between `String` and `Contains` unidirectional.

Saalvage commented on doctest/parts/doctest_fwd.h:
```diff
@@ -576,12 +578,21 @@ class DOCTEST_INTERFACE String
 
 DOCTEST_INTERFACE bool operator==(const String& lhs, const String& rhs);
 DOCTEST_INTERFACE bool operator!=(const String& lhs, const String& rhs);
+DOCTEST_INTERFACE bool operator!=(const String& lhs, const Contains& rhs);
```
Should have `operator==` as well. Just for general utility and maintaining symmetry.

Saalvage commented on doctest/parts/doctest.cpp:
```diff
@@ -1748,7 +1757,27 @@ namespace detail {
         m_threw_as         = false;
         m_exception_type   = exception_type;
-        m_exception_string = exception_string;
+        m_exception_string = String(exception_string);
```
`exception_string` should be made a `String` as well.

Saalvage commented on doctest/parts/doctest_fwd.h:
```diff
@@ -743,7 +754,8 @@ struct DOCTEST_INTERFACE AssertData
     bool        m_threw_as;
     const char* m_exception_type;
-    const char* m_exception_string;
+    String      m_exception_string;
```
I think this should be replaced with a union. Otherwise we have to construct a new `Contains` on every comparison (which includes a string copy), after we've destroyed the one we've been passed in the constructor.

Saalvage commented:
Aside from the few minor tweaks I'd like to see this lgtm!

MFraters commented:
Thanks for the review! I think I have addressed all your issues, I even managed to get the union working :)

I have added some more tests along the lines you suggested, but if you have more in mind, please let me know.

Saalvage commented:
If you can't figure it out just leave it to me, unions can be quite tricky to work with at times and warning suppression is also pretty yucky.

MFraters commented:
Yes, I am happy leave getting the union to you. It work on my computer (linux), but windows is giving a warning/error which is not very helpful and nearly everything I can find on that error is either advice to ignore it or code which upon testing doesn't solve the warning. What also doesn't help is that I can't quickly iterated do not being able to reproduce the issue locally.  

Let me know if you want me to get rid of all the "Attempt to fix windows specific warning about unions" commits, or if there is anything else I can do to move this along.

Saalvage commented:
All good, I'll squash when merging anyways ;)

Saalvage commented:
Don't bother with the test cov ci. This just needs docs and it's ready to go.

Saalvage commented:
Thanks for your contribution!
